docs: Update CHANGELOG.md through version 0.3.0

karfau opened the pull request:
Only referencing #62 for further discussions.

brodybits commented on CHANGELOG.md:
```diff
@@ -24,18 +24,19 @@
 - Remove coveralls.
 - Disable cache in travis.
 - Other refactoring.
+- more details about the transition can be found in #62
```
I think most of the info from #62 should be migrated here.

If we still need any issue references I think any issue references, I think they should be links. Just using issue number reference does not seem to work like issues & PRs.

brodybits commented:
In short, info about regressions and most of the info from issue #62 should be included here.

I think some more info about xmldom vs xmldom-alpha as discussed in issues #7 & #62 should be clarified in CHANGELOG.md as well.

I did add the **bug** label to this. Thanks again!

karfau commented:
Suggestion: It looks like the to issues are complementing each other, but could also understood as duplicate. We would close this one (since it contains less information) and rename the other one so the level of done includes the updated changelog.

What do you think?

brodybits commented:
> What do you think?

👍 (sorry to say, since these updates to CHANGELOG.md do look like a step in the right direction)

karfau commented:
Just realized that this is a PR, not an issue. My plan is to continue on this. So I'm just marking it as a draft for now.

karfau commented:
From my point of view this is done, as far as already published versions are concerned.

I would suggest to introduce (something like) https://github.com/CookPete/auto-changelog to make it easy to update the CHANGELOG in the future.

This is what it could look like with minimal configuration and manual tweaks:
https://github.com/karfau/xmldom/compare/karfau:patch-1...experiment-auto-changelog

I can take care of it directly after this PR, so it can already be applied when releasing 0.4.0

brodybits commented:
I am reviewing this now, my apologies for the delay.

+1 for automating the change log, with some questions coming up

brodybits commented on CHANGELOG.md:
```diff
@@ -1,41 +1,81 @@
+- **BREAKING** Removed component.json (deprecated package manager https://github.com/componentjs/guide)
+- **BREAKING** Moved existing sources into `lib` subdirectory.
+- Update devDependency `proof`
+- Removed `.npmignore` in favor of `files` entry in `package.json`.
```
Is removing `. npmignore ` visible from the npm package?

(I think it is but am not 100% sure. I do recall npm always converting `.gitignore` to `.npmignore`.)

I would maybe turn this around:

```suggestion
- Introduce `files` entry in `package.json` and remove use of `.npmignore`.
```

brodybits commented on CHANGELOG.md:
```diff
@@ -1,41 +1,81 @@
+- **BREAKING** Moved existing sources into `lib` subdirectory.
+- Update devDependency `proof`
+- Removed `.npmignore` in favor of `files` entry in `package.json`.
+
```
remove extra blank line (??)

```suggestion
```

brodybits commented on CHANGELOG.md:
```diff
@@ -1,41 +1,81 @@
+
+**Includes all BREAKING changes introduced in [`xmldom-alpha@v0.1.28`](#0128-only-via-xmldom-alpha-npm-package) and in later commits to master by the original authors.**
+
+- **BREAKING?** [removes the `Object.create` check from the `_extends` method of `dom.js` that added a `__proto__` property](https://github.com/xmldom/xmldom/commit/0be2ae910a8a22c9ec2cac042e04de4c04317d2a#diff-7d1c5d97786fdf9af5446a241d0b6d56L19-L22) ().
+- **BREAKING?** [removed code that added a `__proto__` property](https://github.com/xmldom/xmldom/commit/366159a76a181ce9a0d83f5dc48205686cfaf9cc)
```
I don't really like `?` in changelog.

I would sometimes add "POSSIBLY BREAKING" to commit messages for changes when I am unsure about this.

brodybits commented on CHANGELOG.md:
```diff
@@ -1,41 +1,81 @@
 
-- Removed __proto__ accessor.
-- Appended HTML entities defaults.
+## 0.1.27 (via xmldom and xmldom-alpha npm package)
```
nit:

```suggestion
## 0.1.27 (via xmldom and xmldom-alpha npm packages)
```

brodybits commented on CHANGELOG.md:
```diff
@@ -1,41 +1,81 @@
 
 - Various bug fixes.
-- Don't ask why 0.1.26 and 0.1.25 were skipped. `¯\_(ツ)_/¯`
+- Don't ask why 0.1.26 and 0.1.25 were skipped here, but they have been released only to xmldom npm package on the same day as 0.1.27. `¯\_(ツ)_/¯`
```
How about something like this:

```suggestion
## 0.1.26 - only via xmldom-alpha npm package

- Various bug fixes.

## 0.1.25 - only via xmldom-alpha npm package

- Various bug fixes.
```

brodybits commented on CHANGELOG.md:
```diff
@@ -1,41 +1,81 @@
+- Don't ask why 0.1.26 and 0.1.25 were skipped here, but they have been released only to xmldom npm package on the same day as 0.1.27. `¯\_(ツ)_/¯`
 
-## 0.1.24 (via xmldom-alpha package)
+## 0.1.24 (via xmldom and xmldom-alpha package)
```
nit:
```suggestion
## 0.1.24 - via xmldom and xmldom-alpha packages
```

brodybits reviewed (commented):
I left some nits & minor suggestions. In general:

- I would favor keeping present-tense verbs
- I would avoid parens in the version section headers, to keep difference from generated anchors as minimal as possible (I hope we will never have to deal with xmldom vs xmldom-alpha packages again)

brodybits commented:
I have never seen `auto-changelog` before but am starting to like it. I had seen "conventional commits" before in conventionalcommits.org, with a different tool for updating the changelog and maybe for updating the package version. But it looks like these tools should be able to work together, at least in theory.

> This is what it could look like with minimal configuration and manual tweaks:
> [karfau/xmldom@karfau:patch-1...experiment-auto-changelog](https://github.com/karfau/xmldom/compare/karfau:patch-1...experiment-auto-changelog)

I would probably favor putting this into a PR, which should be easier to review and comment on.

As I said before, I generally have not listed devDependencies and test updates in the change logs, open for discussion.

karfau commented:
Thx for the review.

I completely removed the notion of xmldom vs xmldom-alpha packages from the headers and added that information just below the publish date.

I converted everything (including older entries) to present tense.

Also made sure the order is always consistent without any empty lines between them
- **BREAKING**
- **POSSIBLY BREAKING**
- others

For versions `0.1.25` and `0.1.26` I only added the information we know, and not "Various bugfixes". They have been released on the same day as `0.1.27`, so I don't think a lot happened and I don't care enough about those old versions to try and trace the details.

karfau commented:
> I would probably favor putting this into a PR, which should be easier to review and comment on.

Yes, I just didn't want to invest to much time into it, since I will need to rebase it etc and some details still have to be discussed. I also think it shouldn't block the release of 0.4.0 but it would be nice to already include.

I think the tool just helps to list everything and then it's up to the person publishing the package to decide what information to keep.

karfau commented:
Please feel free to merge when you think it's ready.

brodybits commented on CHANGELOG.md:
```diff
@@ -1,82 +1,142 @@
+
+## Maintainer changes
+
+After the last commit to the original repository <https://github.com/jindw/xmldom> on the 9th of May 2017, the first commit to this fork is from the 19th of December 2019. [The fork has been announced in the original repository on the 2nd of March 2020.](https://github.com/jindw/xmldom/issues/259)
```
Subjective: I am torn at this point whether to refer to this repo as a "fork" or as a new project location.

Assuming we continue to publish into the same `xmldom` package on npm, I would not really favor calling this a "fork".

OTOH if we would start publishing this as multiple packages in the `@xmldom` namespace on npm, which I had already reserved as an org (see issue #54), then I think "fork" would be appropriate.

karfau replied:
what about replacing `this fork` with`<https://github.com/xmldom/xmldom>` so it stays correct even if it should ever move somewhere else.

I think "the announcement of the fork" is still the correct wording for describing what happened back then.

brodybits commented on CHANGELOG.md:
```diff
@@ -1,82 +1,142 @@
+- <https://www.npmjs.com/package/xmldom-alpha>
+- <https://www.npmjs.com/package/xmldom>
+
+The new maintainers do not plan to continue publishing the `xmldom-alpha` package.
```
nit: Not all existing maintainers are "new". Namely @kethinov came from the original `xmldom` project location. And previous maintainers are welcome to join together with us as well. More details are in issue #9.

karfau replied:
So it could just say "The current maintainers ...", right?
Or we could simplify to "There are currently no plans to continue publishing ..."

Do you want to add a sentence about "old and new maintainers" to that section? Feel free to just do that.

brodybits commented on CHANGELOG.md:
```diff
@@ -1,82 +1,142 @@
+
+The new maintainers do not plan to continue publishing the `xmldom-alpha` package.
+
+The new maintainers did not invest time to understand changes that led to the last `xmldom` version [`0.1.27`](#0127) published by the original maintainer, but consider it the basis for their work.
```
nit:

```suggestion
The new maintainer did not invest time to understand changes that led to the last `xmldom` version [`0.1.27`](#0127) published by the original maintainer, but consider it the basis for their work.
```

the "did not invest time" part is very direct, and unfortunately very accurate as well

brodybits replied:
or to turn the knife:

```suggestion
The new maintainer @brodybits did not invest time to understand changes that led to the last `xmldom` version [`0.1.27`](#0127) published by the original maintainer, but consider it the basis for their work.
```

karfau replied:
Well not anybody of the "new maintainers" (which I consider to be the new xmldom org + contributors) took the time to do that, So I would leave the sentence as is.

And in case anybody takes the time, the sentence can be removed.

brodybits commented on CHANGELOG.md:
```diff
@@ -1,82 +1,142 @@
+The new maintainers do not plan to continue publishing the `xmldom-alpha` package.
+
+The new maintainers did not invest time to understand changes that led to the last `xmldom` version [`0.1.27`](#0127) published by the original maintainer, but consider it the basis for their work.
+A time line of all the things that happened from that version until `0.3.0` is available in <https://github.com/xmldom/xmldom/issues/62> and related questions should be asked there.
```
nits:
```suggestion
A timeline of all the changes that happened from that version until `0.3.0` is available in <https://github.com/xmldom/xmldom/issues/62>. Any related questions should be asked there.
```

brodybits commented on CHANGELOG.md:
```diff
@@ -1,82 +1,142 @@
+Published by @jindw on the 9th of May 2017 as
+- `xmldom-alpha@0.1.28`
+
+  - **BREAKING** [regression regarding `&nbsp;`](https://github.com/xmldom/xmldom/issues/57) 
```
nits - ??

```suggestion
- **BREAKING** includes [regression regarding `&nbsp;` (issue #57)](https://github.com/xmldom/xmldom/issues/57) 
```

I am proposing to include a verb of slightly different form (singular) from the other entries, leaving open for discussion.

brodybits commented on CHANGELOG.md:
```diff
@@ -1,82 +1,142 @@
+## 0.1.25
+
+Published on the 18th of Nov 2016 as 
+- `xmldom@0.1.24`
```
???:

```suggestion
- `xmldom@0.1.25`
```

brodybits commented on CHANGELOG.md:
```diff
@@ -1,82 +1,142 @@
 
-- Fixed issue #68, infinite loop on unclosed comment.
-- Added error report for unclosed tag.
+- [Fix infinite loop on unclosed comment.](https://github.com/jindw/xmldom/issues/68)
```
```suggestion
- Fix [infinite loop on unclosed comment (jindw/xmldom#68)](https://github.com/jindw/xmldom/issues/68)
```

(open for discussion)

brodybits reviewed (commented):
I took the liberty to rename this PR (2x) and left what are mostly nits.

One more nit: It looks like we are missing information before version 0.1.18.

I discovered that there is a version 0.1.17 on npm but not in the git history. This would probably be around 2013 or 2014. They may have published 0.1.17 from a special branch or tag in the original repository.

I would probably favor adding a note that details are unknown for versions before 0.1.18.

@karfau I think you should be able to merge this no problem, since it would not be part of a release tag (as a squash commit, please).

brodybits commented:
> > This is what it could look like with minimal configuration and manual tweaks:
> > [karfau/xmldom@karfau:patch-1...experiment-auto-changelog](https://github.com/karfau/xmldom/compare/karfau:patch-1...experiment-auto-changelog)
> 
> I would probably favor putting this into a PR

[...]

> I also think it shouldn't block the release of 0.4.0 but it would be nice to already include.
> 
> I think the tool just helps to list everything and then it's up to the person publishing the package to decide what information to keep.

That makes sense to me.

Maybe a draft PR, that I could use as a starting point for the release of 0.4.0, would make sense. Or maybe not.

It *may* be ideal to set this up to make npm releases from GitHub actions, someday in the future.

karfau commented on CHANGELOG.md:
```diff
@@ -1,82 +1,142 @@
-- Now publishing under xmldom npm package again; retiring xmldom-alpha.
+[Commits](https://github.com/xmldom/xmldom/compare/v0.1.27...0.2.0)
+
+**Includes all BREAKING changes introduced in [`xmldom-alpha@v0.1.28`](#0128) and in later commits to master by the original authors.**
```
Ah, I see this is misleading, I think I fixed this now. I thought I shouldn't duplicate the changes from `0.1.28` here, that's why I linked them. the commits on master are of course what is described below.

karfau commented:
I will merge this on Sunday evening, in case there is no update from your side until then.
But as I said, fell free to just press the button in case you want to go ahead with it earlier.

brodybits commented on CHANGELOG.md:
```diff
@@ -90,3 +147,4 @@
 - Fix: remove existed attribute before setting  (bug introduced in v0.1.5)
 - Fix: index direct access for childNodes and any NodeList collection(not w3c standard)
 - Fix: remove last child bug
+
```
nit: remove extra line from here

```suggestion
```

brodybits commented:
Approved on my part for merge in a squash commit. (Please clean up the squash commit message, and I would appreciate co-author credit.)

brodybits commented:
Merge looks perfect, thanks!

FYI I will not likely respond on GitHub until coming Monday 5th of October.
